### PR TITLE
Add inner hits support to semantic query

### PR DESCRIPTION
**Pull request by Mikep86**
Adds inner hits support to the `semantic` query through a restricted `inner_hits` parameter, which exposes `from` and `size` from the [inner_hits options](https://www.elastic.co/guide/en/elasticsearch/reference/current/inner-hits.html#inner-hits-options). We chose to do this because the other inner hits options don't make sense in the context of the `semantic` query.

The API and response looks like:

```json
PUT test-index
{
    "mappings": {
        "properties": {
            "infer_field": {
                "type": "semantic_text",
                "inference_id": "my-elser-endpoint"
            }
        }
    }
}

PUT test-index/_doc/doc1
{
    "infer_field": ["these are not the droids you're looking for. He's free to go around", "next time on battle bots" ]
}

GET test-index/_search
{
    "query": {
        "semantic": {
            "field": "infer_field",
            "query": "robots",
            "inner_hits": {
                "from": 0,
                "size": 2
            }
        }
    },
    "_source": [ "infer_field.text" ]
}
```

```json
{
    "took": 87,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 1,
            "relation": "eq"
        },
        "max_score": 4.733991,
        "hits": [
            {
                "_index": "test-index",
                "_id": "doc1",
                "_score": 4.733991,
                "_source": {
                    "infer_field": {
                        "text": [
                            "these are not the droids you're looking for. He's free to go around",
                            "next time on battle bots"
                        ]
                    }
                },
                "inner_hits": {
                    "infer_field": {
                        "hits": {
                            "total": {
                                "value": 2,
                                "relation": "eq"
                            },
                            "max_score": 4.733991,
                            "hits": [
                                {
                                    "_index": "test-index",
                                    "_id": "doc1",
                                    "_nested": {
                                        "field": "infer_field.inference.chunks",
                                        "offset": 1
                                    },
                                    "_score": 4.733991,
                                    "_source": {
                                        "text": "next time on battle bots"
                                    }
                                },
                                {
                                    "_index": "test-index",
                                    "_id": "doc1",
                                    "_nested": {
                                        "field": "infer_field.inference.chunks",
                                        "offset": 0
                                    },
                                    "_score": 4.283129,
                                    "_source": {
                                        "text": "these are not the droids you're looking for. He's free to go around"
                                    }
                                }
                            ]
                        }
                    }
                }
            }
        ]
    }
}
```

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InnerChunkBuilder.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        PARSER.declareInt(InnerChunkBuilder::setFrom, SearchSourceBuilder.FROM_FIELD);
+        PARSER.declareInt(InnerChunkBuilder::setSize, SearchSourceBuilder.SIZE_FIELD);
+    }
+
+    private final String name;
+    private int from = DEFAULT_FROM;
+    private int size = DEFAULT_SIZE;
+
+    public InnerChunkBuilder() {
+        // Set hard-coded name value here so that if we change it in the future, the updated value is serialized/deserialized and propagated
```
Comment (kderusso) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InnerChunkBuilder.java:40]: I'm not sure I understand why we need to do this?

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/40_semantic_text_query.yml [hunk 2]
```diff
@@ -90,6 +90,114 @@ setup:
...
+        Content-Type: application/json
+      search:
+        index: test-sparse-index
+        body:
+          query:
+            semantic:
+              field: "inference_field"
+              query: "inference test"
+              inner_chunks: {
+                "size": 1
```
Comment (kderusso) [inline x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/40_semantic_text_query.yml:145]: Can we also add tests for invalid from/size, e.g. -1? 
Could also be worth a test that specifies both size and from, adding a 3rd chunk to validate things got dropped off.

**Review by kderusso — COMMENTED** [review]
Technical approach looks reasonable to me, nice work!

**Review by jimczi — COMMENTED** [review]
+1 for the approach, I left some minor comments.

**Comment by joemcelroy** [comment]
what happens when I query on two semantic_text fields - does it collide under the `inner_hits.chunks` path? We should probably specify a custom name thats unique to the field to avoid collisions.

My first thought is that the retrieved chunks could be positioned more closely to the field source ie `_source.infer_field.chunks` but theres a challenge when also retrieved via `fields`.

The other part to this is I would see a need to describe retrieval of adjacent chunks on matched chunks. These chunks tend to be small (for precision) but not very helpful to the LLM for context. It would be great if we could specify bring back adjacent chunks to the matched chunk. Not asking for this feature now but I think the interface needs to step away from being about configuring inner_hits and more focused around the developers needs for chunking.

**Comment by Mikep86** [comment]
@joemcelroy 

> what happens when I query on two semantic_text fields - does it collide under the inner_hits.chunks path? We should probably specify a custom name thats unique to the field to avoid collisions.

It does indeed! Good callout, I will fix this.

> My first thought is that the retrieved chunks could be positioned more closely to the field source ie _source.infer_field.chunks but theres a challenge when also retrieved via fields.

I don't quite follow here. Do you mean that the nested path `infer_field.inference.chunks` is too verbose?

> Looking at the API interface, I think chunking configuration should step away from being close to configuration of inner_hits and more a developer needs for chunking retrieval.
>
> For example: I would see a need to describe retrieval of adjacent chunks on matched chunks. These chunks tend to be small (for precision) but not very helpful to the LLM for context. It would be great if we could specify bring back adjacent chunks to the matched chunk.
>
> Example: does from parameter make sense in this example?

Agreed that this could be useful, but it's quite the scope expansion. I think that would require implementing a new low-level inner hits retrieval mechanism specifically for retrieving chunks. IMO that's best left as a follow up to this, WDYT?

I think the `from` parameter will always make sense as a pagination mechanism, but how it operates would potentially change with chunk-specific inner hits retrieval.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InnerChunkBuilder.java (hunk 1 above)
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InnerChunkBuilder.java:40]: I did this to implement a hard-coded name that will still be serialized so that if we change it in the future, we have one less BwC problem in a mixed-cluster environment. But it's moot now, based on @joemcelroy's comment the name will have to be dynamic anyways.

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/40_semantic_text_query.yml [hunk 3]
```diff
@@ -356,6 +572,59 @@ setup:
...
+
+  - match: { hits.total.value: 1 }
+  - match: { hits.hits.0._id: "doc_1" }
+  - length: { hits.hits.0._source.inference_field_1.inference.chunks: 2 }
+  - length: { hits.hits.0._source.inference_field_2.inference.chunks: 3 }
+  - match: { hits.hits.0.inner_hits.inference_field_1.hits.total.value: 2 }
+  - length: { hits.hits.0.inner_hits.inference_field_1.hits.hits: 2 }
+  - match: { hits.hits.0.inner_hits.inference_field_2.hits.total.value: 3 }
+  - length: { hits.hits.0.inner_hits.inference_field_2.hits.hits: 3 }
+
```
Comment (Mikep86) [inline x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/40_semantic_text_query.yml:627]: @joemcelroy This test should cover the multiple `semantic_text` field use case

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InnerChunkBuilder.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        return size;
+    }
+
+    public InnerChunkBuilder setSize(int size) {
+        this.size = size;
+        return this;
+    }
+
+    public InnerHitBuilder toInnerHitBuilder() {
+        return new InnerHitBuilder(name).setFrom(from).setSize(size);
```
[the remark this answers is not in the document]
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InnerChunkBuilder.java:86]: Any preference on `fields` vs `_source` filtering? If we use `fields`, we probably shouldn't return `_source` in `inner_hits` at all.

**Comment by joemcelroy** [comment]
Yeh thats fair, we can merge this but in parallel, I would like us to think about API design for configuring how we query and retrieve chunks from semantic_text, so it continues with the "great developer experience" that semantic_text has.

I suggest this PR we should update the request API as a first step, only allowing size to be configured and renamed from `inner_chunks`.  I agree we return the response can be a later step.

Im excited for this change though, thanks for putting it together!

**Comment by Mikep86** [comment]
@joemcelroy 

> could we remove the embeddings aswell, only the chunk text?

💯 , talking with Jim about how we want to do that here: https://github.com/elastic/elasticsearch/pull/111834#discussion_r1715239933

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InnerChunkBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Mikep86) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InnerChunkBuilder.java:86]: Answered my own question. `fields` filtering generates more complex output, so I think we should go with `_source` output.

`fields` filtering:

```json
"inner_hits": {
    "infer_field": {
        "hits": {
            "total": {
                "value": 2,
                "relation": "eq"
            },
            "max_score": 4.733991,
            "hits": [
                {
                    "_index": "test-index",
                    "_id": "doc1",
                    "_nested": {
                        "field": "infer_field.inference.chunks",
                        "offset": 1
                    },
                    "_score": 4.733991,
                    "fields": {
                        "infer_field.inference.chunks": [
                            {
                                "text": [
                                    "next time on battle bots"
                                ]
                            }
                        ]
                    }
                },
                {
                    "_index": "test-index",
                    "_id": "doc1",
                    "_nested": {
                        "field": "infer_field.inference.chunks",
                        "offset": 0
                    },
                    "_score": 4.283129,
                    "fields": {
                        "infer_field.inference.chunks": [
                            {
                                "text": [
                                    "these are not the droids you're looking for. He's free to go around"
                                ]
                            }
                        ]
                    }
                }
            ]
        }
    }
}
```

`_source` filtering:

```
"inner_hits": {
    "infer_field": {
        "hits": {
            "total": {
                "value": 2,
                "relation": "eq"
            },
            "max_score": 4.733991,
            "hits": [
                {
                    "_index": "test-index",
                    "_id": "doc1",
                    "_nested": {
                        "field": "infer_field.inference.chunks",
                        "offset": 1
                    },
                    "_score": 4.733991,
                    "_source": {
                        "text": "next time on battle bots"
                    }
                },
                {
                    "_index": "test-index",
                    "_id": "doc1",
                    "_nested": {
                        "field": "infer_field.inference.chunks",
                        "offset": 0
                    },
                    "_score": 4.283129,
                    "_source": {
                        "text": "these are not the droids you're looking for. He's free to go around"
                    }
                }
            ]
        }
    }
}
```

**Comment by benwtrent** [comment]
My 2 cents is:

 - Either we change the parameter `chunks` to `inner_hits` and place restrictions/set specific defaults on it (easy)
 - Change the response value name to `chunks` (difficult)


Users will always have to read documentation on what the parameter name is. "chunks" vs "parts" vs. "passages" vs. "inner_hits".

**Comment by Mikep86** [comment]
> My 2 cents is:
>
> - Either we change the parameter chunks to inner_hits and place restrictions/set specific defaults on it (easy)
> - Change the response value name to chunks (difficult)

In the spirit of progress over perfection, the first option allows us to merge this change and utilize passage ranking through the `semantic` query now. That makes it the clear choice for the moment IMO.

Given the LOE to make the second option happen, I just don't see it happening without a discovery phase with a lot of stakeholders and multiple PRs. I would likely spend the majority of 8.16 development time on changes related to the search response format, at the cost of other semantic text features we want to include in 8.16.

I also think there's a third option to keep the API as it is now though. `chunks` is basically the request-side API we want, so if we keep it there is one less breaking change to make in the future.

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 5]
```diff
@@ -40,9 +40,209 @@ The `semantic_text` field to perform the query on.
 (Required, string)
 The query text to be searched for on the field.
 
+`chunks`::
+(Optional, object)
+The passage ranking configuration.
```
Comment (carlosdelest) [inline docs/reference/query-dsl/semantic-query.asciidoc:45]:
```suggestion
Retrieves the specific passages that match the query.
```

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 6]
```diff
@@ -40,9 +40,209 @@ The `semantic_text` field to perform the query on.
 (Required, string)
 The query text to be searched for on the field.
 
+`chunks`::
```
Comment (carlosdelest) [inline docs/reference/query-dsl/semantic-query.asciidoc:43]: Naming nit: why not something more specific like passages? (I know we use `chunks` internally, not sure if that was a good idea now that we need to refer to it as such) 🤔

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 7]
```diff
@@ -40,9 +40,209 @@ The `semantic_text` field to perform the query on.
...
+(Optional, object)
+The passage ranking configuration.
+See <<semantic-query-passage-ranking, Passage ranking with the `semantic` query>> for more information.
++
+.Properties of `chunks`
+[%collapsible%open]
+====
+`from`::
+(Optional, integer)
+The offset from the first chunk to fetch.
```
Comment (carlosdelest) [inline docs/reference/query-dsl/semantic-query.asciidoc:53]:
```suggestion
The offset from the first matching passage chunk to fetch.
```

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 8]
```diff
@@ -40,9 +40,209 @@ The `semantic_text` field to perform the query on.
...
+Used to paginate through the chunks.
+Defaults to `0`.
+
+`size`::
+(Optional, integer)
+The maximum number of chunks to return.
```
Comment (carlosdelest) [inline docs/reference/query-dsl/semantic-query.asciidoc:59]:
```suggestion
The maximum number of matching passage chunks to return.
```

**Review by carlosdelest — APPROVED** [review]
LGTM. 💯 

I agree on the approach to use the `inner_hits` name in the request, and keep the inner_hits structure. 

We're already leaking the inner workings of semantic_text - we're just making them convenient to use in terms of ingestion and query. But we're not hiding the details in case users want to dig deeper.

I understand @joemcelroy statement on making it better from the dev perspective. From my side, I think we benefit more of alignment as of now - we're using `inner_hits` config and structure, which is familiar to use and is the inner representation.

We can include the `chunks` parameter later on to make something more aligned to the dev experience that builds on top of this change.

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 9]
```diff
@@ -72,7 +72,10 @@ For example, if you have a document that covers varying topics:
...
 POST my-index/_doc/lake_tahoe
 {
-  "inference_field": [ "Lake Tahoe is the largest alpine lake in North America", "When hiking in the area, please be on alert for bears" ]
+  "inference_field": [
+    "Lake Tahoe is the largest alpine lake in North America",
+    "When hiking in the area, please be on alert for bears"
+  ]
 }
 ------------------------------------------------------------
 // TEST[skip:TBD]
```
Comment (leemthompo) [inline docs/reference/query-dsl/semantic-query.asciidoc:81]: This is a total drive-by nit on docs side here, and no immediate action required, but might be good to start giving more explanatory reasons for skipping docs tests. We're currently auditing docs snippet testing coverage so it would be helpful :)

Even something like `//TEST[skip:Uses ML]` would be more informative.

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 10]
```diff
@@ -25,7 +25,7 @@ GET my-index-000001/_search
   }
 }
 ------------------------------------------------------------
-// TEST[skip:TBD]
+// TEST[skip: Requires inference endpoints]
```
Comment (kderusso) [inline docs/reference/query-dsl/semantic-query.asciidoc:28]: 👍

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 11]
```diff
@@ -40,9 +40,209 @@ The `semantic_text` field to perform the query on.
 (Required, string)
 The query text to be searched for on the field.
 
+`inner_hits`::
+(Optional, object)
+Retrieves the specific passages that match the query.
+See <<semantic-query-passage-ranking, Passage ranking with the `semantic` query>> for more information.
```
Comment (kderusso) [inline docs/reference/query-dsl/semantic-query.asciidoc:46]: Capitalization nitpick 

```suggestion
See <<semantic-query-passage-ranking, passage ranking with the `semantic` query>> for more information.
```

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 12]
```diff
@@ -40,9 +40,209 @@ The `semantic_text` field to perform the query on.
...
++
+.Properties of `inner_hits`
+[%collapsible%open]
```
Comment (kderusso) [inline docs/reference/query-dsl/semantic-query.asciidoc:49]: I don't think we usually collapse query DSL properties, based on looking at a few other pages. It's probably not a huge deal but inconsistent with pages like `knn` and `text_expansion` in the same grouping. Those pages also aren't using the `.Properties` syntax. I'll defer to docs experts on which way is "right" 🙂

File: docs/reference/query-dsl/semantic-query.asciidoc [hunk 13]
```diff
@@ -40,9 +40,209 @@ The `semantic_text` field to perform the query on.
...
+The maximum number of matching passages to return.
+Defaults to `3`.
+====
 
 Refer to <<semantic-search-semantic-text,this tutorial>> to learn more about semantic search using `semantic_text` and `semantic` query.
 
+[discrete]
+[[semantic-query-passage-ranking]]
+==== Passage ranking with the `semantic` query
+The `inner_hits` parameter can be used for _passage ranking_, which allows you to determine which passages in the document best match the query.
```
Comment (kderusso) [inline docs/reference/query-dsl/semantic-query.asciidoc:68]: Would this have been a better place to link to `<<semantic-query-passage-ranking>>`?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java [hunk 14]
```diff
@@ -102,6 +125,21 @@ protected void doWriteTo(StreamOutput out) throws IOException {
         }
         out.writeString(fieldName);
         out.writeString(query);
+        if (out.getTransportVersion().onOrAfter(SEMANTIC_QUERY_INNER_HITS)) {
+            out.writeOptionalWriteable(innerHitBuilder);
+        } else if (innerHitBuilder != null) {
+            throw new IllegalStateException(
+                "Transport version must be at least ["
+                    + SEMANTIC_QUERY_INNER_HITS
```
Comment (kderusso) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java:133]: I think you want this for a friendlier message (this message should never apply in serverless)

```suggestion
                    + SEMANTIC_QUERY_INNER_HITS.toReleaseVersion()
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java [hunk 15]
```diff
@@ -102,6 +125,21 @@ protected void doWriteTo(StreamOutput out) throws IOException {
...
+                    + "] to use [ "
+                    + INNER_HITS_FIELD.getPreferredName()
+                    + "] in ["
+                    + NAME
+                    + "], current transport version is ["
+                    + out.getTransportVersion()
```
Comment (kderusso) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java:139]:
```suggestion
                    + out.getTransportVersion().toReleaseVersion()
```

**Review by kderusso — APPROVED** [review]
Nice work! 👏 A couple of very minor comments surrounding error messages and documentation.

File: docs/reference/query-dsl/semantic-query.asciidoc (hunk 12 above)
Comment (Mikep86) [reply docs/reference/query-dsl/semantic-query.asciidoc:49]: I like the look of the collapsible blocks better personally, that's why I went with it :)

@leemthompo Any guidance you can offer here?

File: docs/reference/query-dsl/semantic-query.asciidoc (hunk 13 above)
Comment (Mikep86) [reply docs/reference/query-dsl/semantic-query.asciidoc:68]: Not sure what you mean here? We usually link to section headings.

Comment (Mikep86) [reply docs/reference/query-dsl/semantic-query.asciidoc:68]: I think I understand now, but I don't see the point of that. This is the first line of the passage ranking tutorial, so linking to the tutorial here would effectively be a link to itself.

Or do I still misunderstand?

Comment (kderusso) [reply docs/reference/query-dsl/semantic-query.asciidoc:68]: 🤦 You're right, I thought the link above was to a different page
